tidy3d/components/data/dataset.py

```diff
@@ -612,24 +637,93 @@ def check_cell_vertex_range(cls, val, values):
-                        log.warning("Unstructured grid contains degenerate cells.")
+                    degenerate_cell_inds = degenerate_cell_inds.union(
+                        np.where(indices[:, i] == indices[:, j])[0]
+                    )
+
+        return degenerate_cell_inds
+
+    @classmethod
+    def _remove_degenerate_cells(cls, cells: CellDataArray):
+        """Remove explicitly degenerate cells if any.
+        That is, cells that use the same point indices for their different vertices.
+        """
+        degenerate_cells = cls._find_degenerate_cells(cells=cells)
+        if len(degenerate_cells) > 0:
+            data = np.delete(cells.values, list(degenerate_cells), axis=0)
+            cell_index = np.delete(cells.cell_index.values, list(degenerate_cells))
+            return CellDataArray(
+                data=data, coords=dict(cell_index=cell_index, vertex_index=cells.vertex_index)
+            )
+        return cells
+
+    @classmethod
+    def _remove_unused_points(
+        cls, points: PointDataArray, values: IndexedDataArray, cells: CellDataArray
+    ):
+        """Remove explicitly degenerate cells if any.
+        That is, cells that use the same point indices for their different vertices.
+        """
+
+        used_indices = np.unique(cells.values.ravel())
+        num_points = len(points)
+
+        if len(used_indices) != num_points or np.any(np.diff(used_indices) != 1):
+            min_index = np.min(used_indices)
+            map_len = np.max(used_indices) - min_index + 1
+            index_map = np.zeros(map_len)
+            index_map[used_indices - min_index] = np.arange(len(used_indices))
+
+            cells = CellDataArray(data=index_map[cells.data - min_index], coords=cells.coords)
+            points = points[used_indices, :]
+            values = values[used_indices]
+
+        return points, values, cells
+
+    def clean(self, remove_degenerate_cells=True, remove_unused_points=True):
+        """Remove degenerate cells and/or unused points."""
+        if remove_degenerate_cells:
+            cells = self._remove_degenerate_cells(cells=self.cells)
+        else:
+            cells = self.cells
+
+        if remove_unused_points:
+            points, values, cells = self._remove_unused_points(self.points, self.values, cells)
+        else:
+            points = self.points
+            values = self.values
+
+        return self.updated_copy(points=points, values=values, cells=cells)
+
+    @pd.validator("cells", always=True)
+    def warn_degenerate_cells(cls, val):
+        """Check that cell connections does not have duplicate points."""
+        degenerate_cells = cls._find_degenerate_cells(val)
+        num_degenerate_cells = len(degenerate_cells)
+        if num_degenerate_cells > 0:
+            log.warning(
+                f"Unstructured grid contains {num_degenerate_cells} degenerate cell(s). "
+                "Consider applying 'clean()'."
```

Comment:
can you elaborate here? where is `clean()` and how would the user specifically call it? (at least add a code snippet, like `dataset = dataset.clean()` or something?

Same comment below